util/reflect.go

```diff
@@ -367,61 +413,368 @@ type NodeInfo struct {
+// to the supplied root. If for a list node is missing, all values in the list
+// are returned. If the key is partial, all nodes matching the values present
+// in the key are returned.
+// If the root is the tree root, the path may be absolute.
+// GetNodes returns an error if the path is not found in the tree, or an element
+// along the path is nil.
+func GetNodes(schema *yang.Entry, root interface{}, path *gpb.Path) ([]interface{}, []*yang.Entry, error) {
+	return getNodesInternal(schema, root, path)
+}
+
+// getNodesInternal is the internal implementation of GetNode. In addition to
+// GetNode functionality, it can accept non GoStruct types e.g. map for a keyed
+// list, or a leaf.
+func getNodesInternal(schema *yang.Entry, root interface{}, path *gpb.Path) ([]interface{}, []*yang.Entry, error) {
+	if IsValueNil(root) {
+		ResetIndent()
+		return nil, nil, nil
+	}
+	if len(path.GetElem()) == 0 {
+		ResetIndent()
+		return []interface{}{root}, []*yang.Entry{schema}, nil
+	}
+	if schema == nil {
+		return nil, nil, fmt.Errorf("nil schema for data element type %T, remaining path %v", root, path)
+	}
+	// Strip off the absolute path prefix since the relative and absolute paths
+	// are assumed to be equal.
+	if path.GetElem()[0].GetName() == "" {
+		path.Elem = path.GetElem()[1:]
+	}
+
+	Indent()
+	DbgPrint("GetNode next path %v, value %v", path.GetElem()[0], ValueStr(root))
+
+	switch {
+	case schema.IsContainer() || (schema.IsList() && IsTypeStructPtr(reflect.TypeOf(root))):
+		// Either a container or list schema with struct data node (which could
+		// be an element of a list).
+		return getNodesContainer(schema, root, path)
+	case schema.IsList():
+		// A list schema with the list data node. Must find the element selected
+		// by the path.
+		return getNodesList(schema, root, path)
+	}
+
+	return nil, nil, fmt.Errorf("bad schema type for %s, struct type %T", schema.Name, root)
+}
+
+// getNodesContainer traverses the container root, which must be a struct ptr
+// type and matches each field against the first path element in path. If a
+// field matches, it recurses into that field with the remaining path.
+func getNodesContainer(schema *yang.Entry, root interface{}, path *gpb.Path) ([]interface{}, []*yang.Entry, error) {
+	DbgPrint("getNodesContainer: schema %s, next path %v, value %v", schema.Name, path.GetElem()[0], ValueStr(root))
+
+	rv := reflect.ValueOf(root)
+	if !IsValueStructPtr(rv) {
+		return nil, nil, fmt.Errorf("getNodesContainer: root has type %T, expect struct ptr", root)
+	}
+
+	v := rv.Elem()
+
+	for i := 0; i < v.NumField(); i++ {
+		f := v.Field(i)
+		ft := v.Type().Field(i)
+		cschema, err := FieldSchema(schema, ft)
+		if err != nil {
+			return nil, nil, fmt.Errorf("error for schema for type %T, field name %s: %s", root, ft.Name, err)
+		}
+		if cschema == nil {
+			return nil, nil, fmt.Errorf("could not find schema for type %T, field name %s", root, ft.Name)
+		}
+		cschema, err = ResolveLeafRef(cschema)
```

Comment:
It's counter-intuitive to me that you call this here even if the field isn't a leafref -- isn't the pattern of:

```
switch {
   case cschema == nil: 
     // return err
   case isLeafref(cschema):
     var err error
     cschema, err = ResolveLeafRef(cschema)
 }
```

make more sense?

Reply:
The ResolveLeafRef function description states "It returns schema itself if schema is not a leaf-ref." but I agree this is a bit confusing because at first read it implies we think we have a leafref when calling the function. 
At the same time I don't want to make the caller check if it's a leafref every time - I could maybe rename this to ResolveIfLeafRef, or otherwise what would you suggest?

Reply:
Changed to ResolveIfLeafRef per offline discussion.